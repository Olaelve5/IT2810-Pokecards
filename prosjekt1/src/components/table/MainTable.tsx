import {useState} from 'react';
import Row from './Row';
import { PokemonType } from '../../types/Pokemon';
import { useDataContext } from '../../contexts/DataContext';
import classes from '../../styles/table/MainTable.module.css';
import PageButtons from '../buttons/PageButtons';
import MainFilter from './filter/MainFilter';
import FavoritesToggle from '../buttons/FavoritesToggle';

const MainTable = () => {
  const { tablePokemons, totalPokemonCount, setTotalPokemonCount } = useDataContext();
  const [page, setPage] = useState<number>(1);

  return (
    <div className={classes.tableContainer}>
      <MainFilter page={page} setPage={setPage} setTotalPokemonCount={setTotalPokemonCount}/>
      <FavoritesToggle page={page} setPage={setPage} totalPokemonCount={totalPokemonCount} setTotalPokemonCount={setTotalPokemonCount}/>
      <div className={classes.grid}>
        {tablePokemons && tablePokemons.length === 0 && (
          <div className={classes.noResults}>No pokemon found</div>
        )}

        {tablePokemons?.map((pokemon: PokemonType) => (
          <div key={pokemon.number} className={classes.cell}>
            <Row {...pokemon} />
          </div>
        ))}
      </div>
      <PageButtons page={page} setPage={setPage} totalPokemonCount={totalPokemonCount}/>
    </div>
  );
};

export default MainTable;
