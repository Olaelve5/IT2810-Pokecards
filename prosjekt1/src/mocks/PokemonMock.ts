const pokemon1 = {
  number: 1,
  name: 'Beedrill',
  images: {
    small: 'https://example.com/small.png',
    large: 'https://example.com/large.png',
  },
};

const pokemon2 = {
  number: 2,
  name: 'Butterfree',
  images: {
    small: 'https://example.com/small.png',
    large: 'https://example.com/large.png',
  },
};

const pokemon3 = {
  number: 3,
  name: 'Pikachu',
  images: {
    small: 'https://example.com/small.png',
    large: 'https://example.com/large.png',
  },
};

export const tablePokemonsMock = [pokemon1, pokemon2, pokemon3];
