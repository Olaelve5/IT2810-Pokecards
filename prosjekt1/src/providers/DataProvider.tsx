import { useEffect, useState } from "react";
import { DataContext } from "../contexts/DataContext";
import { PokemonType } from "../types/Pokemon";

const DataProvider = ({ children }: { children: React.ReactNode }) => {
    const [activePokemon, setActivePokemon] = useState<PokemonType | null>(null);
    const [tablePokemons, setTablePokemons] = useState<PokemonType[]>([]);
    const [showFavorites, setShowFavorites] = useState<boolean>(false);

    useEffect(() => {
        if (!activePokemon && tablePokemons.length > 0) {
            setActivePokemon(tablePokemons[0]);
        }
    }, [activePokemon, tablePokemons]);

    return (
        <DataContext.Provider value={{ 
            activePokemon, setActivePokemon, 
            tablePokemons, setTablePokemons,
            showFavorites, setShowFavorites
            }}>
            {children}
        </DataContext.Provider>
    );
};

export default DataProvider;

