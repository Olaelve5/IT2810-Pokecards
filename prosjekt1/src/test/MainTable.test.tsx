import MainTable from '../components/table/MainTable';
import { tablePokemonsMock } from '../mocks/PokemonMock';
import { render, screen, waitFor } from '@testing-library/react';
import DataProvider from '../providers/DataProvider';
import { QueryClientProvider, QueryClient } from '@tanstack/react-query';
import userEvent from '@testing-library/user-event';

const setActivePokemon = vi.fn();

// Mocking the DataContext
vi.mock('../contexts/DataContext', async () => {
  const actual = await import('../contexts/DataContext');
  return {
    ...actual,
    useDataContext: () => ({
      activePokemon: tablePokemonsMock[0],
      tablePokemons: tablePokemonsMock,
      setActivePokemon,
      setTablePokemons: vi.fn(),
      setTotalPokemonCount: vi.fn(),
      setGlobalTotalPokemonCount: vi.fn(),
    }),
  };
});

// Test if the MainTable component renders the correct table and updates activePokemon on click
describe('MainTable', () => {
  let queryClient: QueryClient;

  beforeEach(() => {
    queryClient = new QueryClient();
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  it('renders table and updates displayName on click', async () => {
    const mainTable = render(
      <QueryClientProvider client={queryClient}>
        <DataProvider>
          <MainTable />
        </DataProvider>
      </QueryClientProvider>,
    );
    expect(mainTable).toMatchSnapshot();

    userEvent.click(screen.getByText(tablePokemonsMock[2].name));

    await waitFor(() => {
      expect(setActivePokemon).toHaveBeenCalledWith(tablePokemonsMock[2]);
    });
  });
});
