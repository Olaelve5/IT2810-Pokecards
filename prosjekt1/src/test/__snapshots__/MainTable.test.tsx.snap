// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MainTable > renders table and updates displayName on click 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_tableContainer_bf6e8c"
      >
        <div
          class="_container_5b5f10"
        >
          <h3
            class="_title_5b5f10"
          >
            Find cards
          </h3>
          <div
            class="_filterContainer_5b5f10"
          >
            <div
              class="_container_d50bc9"
            >
              <p>
                Name
              </p>
              <input
                class="_input_d50bc9"
                placeholder="Ex: Pikachu"
                type="text"
                value=""
              />
            </div>
            <div
              class="_container_d1f101"
            >
              <p>
                Type 
              </p>
              <select
                class="_select_d1f101"
              >
                <option
                  value=""
                >
                  All
                </option>
                <option
                  value="Colorless"
                >
                  Colorless
                </option>
                <option
                  value="Fighting"
                >
                  Fighting
                </option>
                <option
                  value="Fire"
                >
                  Fire
                </option>
                <option
                  value="Grass"
                >
                  Grass
                </option>
                <option
                  value="Lightning"
                >
                  Lightning
                </option>
                <option
                  value="Psychic"
                >
                  Psychic
                </option>
                <option
                  value="Water"
                >
                  Water
                </option>
              </select>
            </div>
            <div
              class="_container_7723c2"
            >
              <p>
                Sort by
              </p>
              <div
                class="_buttonContainer_7723c2"
              >
                <select>
                  <option
                    value="ID"
                  >
                    ID
                  </option>
                  <option
                    value="Name"
                  >
                    Name
                  </option>
                  <option
                    value="HP"
                  >
                    HP
                  </option>
                </select>
                <button>
                  <svg
                    class="tabler-icon tabler-icon-sort-ascending "
                    fill="none"
                    height="20"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="20"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M4 6l7 0"
                    />
                    <path
                      d="M4 12l7 0"
                    />
                    <path
                      d="M4 18l9 0"
                    />
                    <path
                      d="M15 9l3 -3l3 3"
                    />
                    <path
                      d="M18 6l0 12"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          class="_container_daa855"
        >
          <button>
            Show only favorites
          </button>
        </div>
        <div
          class="_grid_bf6e8c"
        >
          <div
            class="_cell_bf6e8c"
          >
            <div
              class="_container_8796b6"
            >
              <div
                class="_nameIdContainer_8796b6"
              >
                <h4>
                  Beedrill
                </h4>
                <p>
                  #
                  1
                </p>
              </div>
              <img
                alt="Beedrill"
                class="_image_8796b6"
                src="https://example.com/small.png"
              />
            </div>
          </div>
          <div
            class="_cell_bf6e8c"
          >
            <div
              class="_container_8796b6"
            >
              <div
                class="_nameIdContainer_8796b6"
              >
                <h4>
                  Butterfree
                </h4>
                <p>
                  #
                  2
                </p>
              </div>
              <img
                alt="Butterfree"
                class="_image_8796b6"
                src="https://example.com/small.png"
              />
            </div>
          </div>
          <div
            class="_cell_bf6e8c"
          >
            <div
              class="_container_8796b6"
            >
              <div
                class="_nameIdContainer_8796b6"
              >
                <h4>
                  Pikachu
                </h4>
                <p>
                  #
                  3
                </p>
              </div>
              <img
                alt="Pikachu"
                class="_image_8796b6"
                src="https://example.com/small.png"
              />
            </div>
          </div>
        </div>
        <div
          class="_container_628c4c"
        >
          <button
            class="_button_628c4c"
            disabled=""
          >
            Previous
          </button>
          <div
            class="_textContainer_628c4c"
          >
            <span>
              1
            </span>
            <span>
              / 
              NaN
            </span>
          </div>
          <button
            class="_button_628c4c"
          >
            Next
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_tableContainer_bf6e8c"
    >
      <div
        class="_container_5b5f10"
      >
        <h3
          class="_title_5b5f10"
        >
          Find cards
        </h3>
        <div
          class="_filterContainer_5b5f10"
        >
          <div
            class="_container_d50bc9"
          >
            <p>
              Name
            </p>
            <input
              class="_input_d50bc9"
              placeholder="Ex: Pikachu"
              type="text"
              value=""
            />
          </div>
          <div
            class="_container_d1f101"
          >
            <p>
              Type 
            </p>
            <select
              class="_select_d1f101"
            >
              <option
                value=""
              >
                All
              </option>
              <option
                value="Colorless"
              >
                Colorless
              </option>
              <option
                value="Fighting"
              >
                Fighting
              </option>
              <option
                value="Fire"
              >
                Fire
              </option>
              <option
                value="Grass"
              >
                Grass
              </option>
              <option
                value="Lightning"
              >
                Lightning
              </option>
              <option
                value="Psychic"
              >
                Psychic
              </option>
              <option
                value="Water"
              >
                Water
              </option>
            </select>
          </div>
          <div
            class="_container_7723c2"
          >
            <p>
              Sort by
            </p>
            <div
              class="_buttonContainer_7723c2"
            >
              <select>
                <option
                  value="ID"
                >
                  ID
                </option>
                <option
                  value="Name"
                >
                  Name
                </option>
                <option
                  value="HP"
                >
                  HP
                </option>
              </select>
              <button>
                <svg
                  class="tabler-icon tabler-icon-sort-ascending "
                  fill="none"
                  height="20"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M4 6l7 0"
                  />
                  <path
                    d="M4 12l7 0"
                  />
                  <path
                    d="M4 18l9 0"
                  />
                  <path
                    d="M15 9l3 -3l3 3"
                  />
                  <path
                    d="M18 6l0 12"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="_container_daa855"
      >
        <button>
          Show only favorites
        </button>
      </div>
      <div
        class="_grid_bf6e8c"
      >
        <div
          class="_cell_bf6e8c"
        >
          <div
            class="_container_8796b6"
          >
            <div
              class="_nameIdContainer_8796b6"
            >
              <h4>
                Beedrill
              </h4>
              <p>
                #
                1
              </p>
            </div>
            <img
              alt="Beedrill"
              class="_image_8796b6"
              src="https://example.com/small.png"
            />
          </div>
        </div>
        <div
          class="_cell_bf6e8c"
        >
          <div
            class="_container_8796b6"
          >
            <div
              class="_nameIdContainer_8796b6"
            >
              <h4>
                Butterfree
              </h4>
              <p>
                #
                2
              </p>
            </div>
            <img
              alt="Butterfree"
              class="_image_8796b6"
              src="https://example.com/small.png"
            />
          </div>
        </div>
        <div
          class="_cell_bf6e8c"
        >
          <div
            class="_container_8796b6"
          >
            <div
              class="_nameIdContainer_8796b6"
            >
              <h4>
                Pikachu
              </h4>
              <p>
                #
                3
              </p>
            </div>
            <img
              alt="Pikachu"
              class="_image_8796b6"
              src="https://example.com/small.png"
            />
          </div>
        </div>
      </div>
      <div
        class="_container_628c4c"
      >
        <button
          class="_button_628c4c"
          disabled=""
        >
          Previous
        </button>
        <div
          class="_textContainer_628c4c"
        >
          <span>
            1
          </span>
          <span>
            / 
            NaN
          </span>
        </div>
        <button
          class="_button_628c4c"
        >
          Next
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
